### ISSUE TITLE
hsc2hs invocation picks up too many flags from pkgdb

### ISSUE THREAD

**Issue by ezyang** [haskell/cabal#2971]
Cabal computes the transitive dependency of packages, and uses `include-paths` and other fields to determine what to pass to `hsc2hs`. But it grabs too many packages: it does the entire set of all packages that ANY component in the package might need, rather than the component that is being built. Test case attached. I'll push a comment to the code that needs to be fixed too.
[test.zip](https://github.com/haskell/cabal/files/69071/test.zip)

To run the test, first build p and q, and note that it fails. Now touch `/tmp/asdf.h`, and observe that it now succeeds. p adds `/tmp` to import path, but `q`'s buildable executable does not refer to p and should not see it.

**Comment by ezyang** [comment]
When #3022 is merged this should only take a local change to fix.

**Comment by ezyang** [comment]
Actually, there is a bit more refactoring to do. The problem is the PackageIndex we store in `packageDependsIndex` doesn't contain the INTERNAL components we have, so it's inconvenient to ask for the closure of a given subcomponent; it will refer to things that are not in the index. Actually I think this reveals a bug, where if this won't build:

```
name:                PreProcessInternal
version:             0.1.0.0
license:             BSD3
author:              Edward Z. Yang
maintainer:          ezyang@cs.stanford.edu
build-type:          Simple
cabal-version:       >=1.10

library
  build-depends:       base
  include-dirs:        include
  install-includes:    asdf.h
  default-language:    Haskell2010

executable exe
  main-is: Main.hs
  build-depends:       base, PreProcessInternal
```

(where `Main.hsc` includes `asdf.h`).

**Comment by ezyang** [comment]
I accidentally fixed this on my way for another 2.0 blocker, so a patch is on the way.
